### ISSUE TITLE
[Bug] Fused multi-layer kernel: last layer's add_inout write to k_cache is not flushed back to host

### ISSUE THREAD

**Issue by wangqin1723-max** [hw-native-sys/pypto#1287]
### Component

Runtime / device-host sync (RT2)

### Description

When a fused multi-layer kernel uses `pl.unroll(N)` to repeatedly write to the same `pl.Tensor` parameter (which the orchestration codegen emits as `add_inout` in `pto2_rt_submit_aiv_task` calls), the **last iteration's writes do not propagate back to the host buffer** after the kernel call returns. The kernel computation succeeds on the device side (downstream attention reads see the just-written K and produce correct outputs), but the host-visible buffer is byte-identical before and after the call.

### Steps to Reproduce

In `hw-native-sys/pypto-lib`, branch `feat/hf-compare-framework`, file `examples/models/qwen3/14b/qwen3_14b_decode_full.py`:

1. Confirm line 162 reads `for layer_idx in pl.unroll(num_layers):`
2. Run a single decode step with the 3-layer model:

```bash
task-submit --device <free> --run "python llm/examples/qwen3_14b_npu_generate.py \
    --model-dir <Qwen3-14B-3L-path> \
    --prompt '你好' --platform a2a3 \
    --max-seq-len 128 --max-new-tokens 1"
```

3. In `llm/core/pypto_executor.py` `run_decode`, snapshot `k_cache` before/after the kernel call:

```python
k_cache_before = k_cache.clone()
compiled.decode(..., k_cache, ..., out, ...)
for li in range(num_layers):
    base = li * layer_cache_rows
    modified = not torch.equal(
        k_cache_before[base:base + layer_cache_rows],
        k_cache[base:base + layer_cache_rows],
    )
    print(f"L{li}: modified={modified}")
```

### Expected Behavior

For `num_layers=3`, all three layers' K cache regions should be modified (every layer's `pl.assemble` writes to `k_cache` at slot offset `layer_idx * layer_cache_rows + ...`).

### Actual Behavior

Only L0 and L1 are modified on host:

```
[probe-decode-perlayer] L0:max_abs=8.875e+00 | L1:max_abs=1.662e+01 | L2:max_abs=2.912e+01
[probe-decode-kvwrite]  L0:modified=True,diff_max=109 | L1:modified=True,diff_max=31 | L2:modified=False,diff_max=0
[probe-decode]          out: finite=True max_abs=29.12   ← out is correct!
```

L2's K cache region (rows `[2 * layer_cache_rows, 3 * layer_cache_rows)`) is byte-identical to the snapshot before the kernel call.

Critically, **the kernel's internal compute is correct**:
- L2's `hidden` output (29.12) is finite and shows the expected residual stream growth (8.87 → 16.62 → 29.12)
- For L2's attention to produce a finite output, it must have read the K it just wrote at slot 1 (other positions are from prefill)
- So **the K write does land on the device side**, but never gets flushed back to host

For `num_layers=4`, both L2 and L3 are modified=False; L1 develops NaN as a secondary effect on subsequent decode steps because the cross-call KV cache reuse picks up stale data/w00956228/newpto.

### Codegen Evidence

In `build_output/Qwen3Decode_*/orchestration/qwen3_decode.cpp`:

```cpp
// L0: first K write — uses add_output (presumably always flushed)
params_t6.add_output(ext_k_cache);

// L1, L2, L3: subsequent writes — all use add_inout
params_t26.add_inout(ext_k_cache);   // L1
params_t46.add_inout(ext_k_cache);   // L2  ← submitted, but not flushed
params_t66.add_inout(ext_k_cache);   // L3  ← submitted, but not flushed (4L only)
```

All four `pto2_rt_submit_aiv_task(N, params_tN)` calls are emitted. The corresponding `qwen3_decode_incore_*.cpp` kernel files all contain `TSTORE` instructions targeting the K cache pointer with the correct per-layer offsets (verified by inspecting `v77 = (slot_block * num_kv_heads + ki) * BLOCK_SIZE + slot_offset + layer_offset_const`).

### Numerical Summary

| Config | L0 | L1 | L2 | L3 |
|---|---|---|---|---|
| 3L unroll | modified=True ✓ | modified=True ✓ | **modified=False ✗** | — |
| 4L unroll | modified=True ✓ | modified=True ⚠ NaN | **modified=False ✗** | **modified=False ✗** |

### Working Hypothesis

For an `add_inout` tensor written multiple times within a single fused kernel call, RT2's host-flush logic only flushes the first ~N-1 writes (where the count of intermediate writes that get RAW-followed by reads in the same call). The very last write to that tensor — even if subsequent decode-step kernel calls will read it as input — appears to be elided from the host writeback path.

### Why This Matters

This makes fused multi-layer decode kernels produce silently wrong KV cache on host, breaking any usage that requires KV cache reuse across decode steps (i.e., all autoregressive generation). A 3-layer model produces gibberish text; a 4-layer model produces NaN logits.

The single-layer kernel (`qwen3_14b_decode.py`) called N times from python (per-layer loop in the executor) works correctly because each call only has 1 inout write. The fused multi-layer version is the target architecture for performance, but is unusable until this is fixed.

### Git Commit ID

repo: `hw-native-sys/pypto-lib`, branch `feat/hf-compare-framework`, kernel file `examples/models/qwen3/14b/qwen3_14b_decode_full.py`; pypto commit dd8d521e or later

### NPU Kind

Ascend 910B

### Host Platform

Linux (aarch64)

**Comment by wangqin1723-max** [comment]
Closing — root cause is in pypto-lib, not in the pypto runtime.

I instrumented `validate_runtime_impl` to read each tensor's dev_ptr at multiple offsets before the bulk copy-back: the K-cache dev_ptr is allocated at the full host size and every layer's TSTORE actually lands on GM and gets copied back. So RT2 does not elide any writes — the original "last add_inout is not flushed" hypothesis was wrong.

Real cause: `pypto-lib/examples/models/qwen3/14b/qwen3_14b_decode_full.py` uses `BLOCK_SIZE = SEQ_TILE = 64` as both
1. the attention compute tile size, and
2. the KV-cache page size used to compute `layer_cache_rows` and per-slot offsets.

But the host runtime config (`qwen3_14b_npu_generate.py`) uses `page_size = 256`. The fused multi-layer kernel computes `layer_cache_rows = 1024` while the host KV cache is laid out at `2048` rows/layer, so every layer's writes land in the wrong host layer region (L1 K → host L0; L2 K → host L1; host L2 → never written). That is exactly the `L0=True, L1=True, L2=False` pattern observed.

The single-layer decode worked because, without an inter-layer offset, the kernel's layout is self-consistent within one host-sliced region. Only the fused kernel exposes the implicit "kernel BLOCK_SIZE must equal host page_size" coupling.

Fix lives in pypto-lib (separate PR): make `page_size` an explicit parameter of `build_qwen3_decode_program` (decoupled from `SEQ_TILE`) and have the executor pass `model.runtime.page_size`. After the fix, all three layers report `modified=True` with sensible diffs in `[probe-decode-kvwrite]`.

(The L2 hidden still goes NaN after the fix, but that is an unrelated compute-side bug — KV-cache plumbing is now correct.)
